_posts/grant_guo/2018-03-22-kf-etl.md

```diff
@@ -0,0 +1,349 @@
+* Supports both proto2 and proto3. The ETL uses `proto2` because some fields of the source data are optional and only `proto2` supports the keywords `optional` and `required`. 
+* Supports SparkSQL
+* Supports converting to and from JSON. The ETL uses this to directly generate JSON strings.
+* Supports generating case classes for all messages. Applying case classes to Spark Dataset is intuitive, so that no extra [Encoder](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Encoder)s are needed.
+* Supports introspection of generated case classes 
+
+### Context
+
+`Context` holds globally shared resources like `SparkSession`, `PostgresQL`, and `Elasticsearch` and is defined like this:
+
+```
+object Context extends ContextTrait with ClasspathURLEnabler{
+  lazy val (hdfs, rootPath) = getHDFS()
+  lazy val sparkSession = getSparkSession()
+  lazy val postgresql = getPostgresql()
+  lazy val esClient = getESClient()
+  
+  ...
+}
+```
+
+### Processor
+
+`Processor` is the key abstraction in the ETL and is defined like this:
+```
+trait Processor[I, O] extends Function1[I, O]{
+  override def apply(input: I): O = {
+    process(input)
+  }
+  def process(input:I): O
+}
+```
+
+The `Processor` is derived from Scala's [Function1[-T1, +R]](http://www.scala-lang.org/api/2.11.12/#scala.Function1) which basically means that a `Processor` is just a function. Any concrete `Processor` fills its logic by overriding the function `process(input:I):O`. Technically, a `Processor` could be implemented in any direct or indirect way. However, the reference implementations follow this convention:
+
+> A Processor is composed of source, transform, sink and output. A transform can be composed of Steps. The ProcessorContext and  StepContext hold shared resources for a Processor and a Step respectively.
+
+`Source`, `transform`, `sink` and `output` are types, but the ETL doesn't define the particular base interface respectively. Instead, they are all derived from Scala [Function1[-T1, +R]](http://www.scala-lang.org/api/2.11.12/#scala.Function1). This means that a `Processor` can be implemented in a very consistent and simple way like this:
+
+```
+class FileCentricProcessor(context: FileCentricContext,
+                           source: Repository => DatasetsFromDBTables,
+                           transform: DatasetsFromDBTables => Dataset[FileCentric],
+                           sink: Dataset[FileCentric] => Unit,
+                           output: Unit => (String,Repository)) extends Processor[Repository, (String,Repository)]{
+
+  override def process(input: Repository):(String,Repository) = {
+    source.andThen(transform).andThen(sink).andThen(output)(input)
+  }
+
+}
+``` 
+
+The key point is the usage of the `andThen` function defined in Scala [Function1[-T1, +R]](http://www.scala-lang.org/api/2.11.12/#scala.Function1). `andThen` allows  function calls to be changed together sequentially rather than nested together. 
```

Comment:
The key point of what?
Do you just want to say something like: "A very useful function for our ETL processing is Scala's `andThen` function, which allows function calls ..."